core/services/functions/orm.go

```diff
@@ -186,11 +189,42 @@ func (o orm) FindOldestEntriesByState(state RequestState, limit uint32, qopts ..
 	return requests, nil
 }
 
-func (o orm) FindById(requestID RequestID, qopts ...pg.QOpt) (*Request, error) {
+func (o *orm) FindById(requestID RequestID, qopts ...pg.QOpt) (*Request, error) {
 	var request Request
 	stmt := fmt.Sprintf(`SELECT %s FROM ocr2dr_requests WHERE request_id=$1 AND contract_address=$2;`, requestFields)
 	if err := o.q.WithOpts(qopts...).Get(&request, stmt, requestID, o.contractAddress); err != nil {
 		return nil, err
 	}
 	return &request, nil
 }
+
+func (o *orm) PruneOldestRequests(maxStoredRequests uint32, batchSize uint32, qopts ...pg.QOpt) (total uint32, pruned uint32, err error) {
+	err = o.q.WithOpts(qopts...).Transaction(func(tx pg.Queryer) error {
+		if err2 := tx.Get(&total, "SELECT COUNT(*) FROM ocr2dr_requests WHERE contract_address=$1", o.contractAddress); err2 != nil {
+			return errors.Wrap(err, "failed to get request count")
+		}
+
+		if total <= maxStoredRequests {
+			pruned = 0
+			return nil
+		}
+
+		pruneLimit := total - maxStoredRequests
+		if pruneLimit > batchSize {
+			pruneLimit = batchSize
+		}
+
+		with := `WITH ids AS (SELECT request_id FROM ocr2dr_requests WHERE contract_address = $1 ORDER BY received_at LIMIT $2)`
```

Comment:
BTREE only works well if the thing it's indexing is somewhat unique... so for example, if you have a single contract address per row then it will be perfect... if there are only a handful of rows per contract address, still okay.  But if there are only a small number of contract addresses each with a large number of rows it won't help at all.  HASH is the opposite and will help the most if you have a small number of contract addresses, but still should work if you have lots of contract addresses each with a large number of rows.